ibis/common/grounds.py

```diff
@@ -9,7 +9,7 @@
 
 from .. import util
 from .caching import WeakCache
-from .validators import Optional, Validator
+from .validators import Optional, Promoter, Validator
```

Comment:
Can we call this `SlottedProperty` or `ImmutableProperty` (and change the corresponding decorator? It doesn't seem clear what the relationship between promotion and caching immutable frequently accessed properties is.